Shelf crud

siripr4 opened the pull request:
Implement CRUD APIs and tests on the `shelf` entity
```json
{
    "id": "0305335d-78b9-4649-a7e8-09aaac374127",
    "label": "2A",
    "section": "A",
    "level": "2",
    "shelfBlockId": "0bbe5f5c-d8ea-4bc8-ad06-2ada61a23048"
}
```

siripr4 committed:
refactor: rename shelfService -> ShelfBlockService

siripr4 committed:
feat: shelf crud and tests

olttwa commented on warehouse-management-service/internal/handler/handler.go:
```diff
@@ -469,7 +479,226 @@ func (h *handler) DeleteShelfBlock(w http.ResponseWriter, r *http.Request) {
+	}
 
+	shelf, err := h.shelfService.GetShelfById(r.Context(), shelfId)
+	switch err {
```
I understand that an error here can be 1 of 3 values. However, that still doesn't justify using a switch statement for that. I'd write this like because of 2 reasons:
1. Convention of error checking in Golang
2. Putting non-erroneous statements like http.StatusOK response in same indentation as the flow of program improves readability.

```golang
warehouse, err := h.warehouseService.GetWarehouseById(r.Context(), warehouseId)
	if err != nil {
		if err == wms.WarehouseDoesNotExist {
			h.logger.Log(log.Error, err)
			h.response(w, http.StatusNotFound, api.GetWarehouseResponse{Error: fmt.Sprintf(
				"failed to get, warehouse: %s does not exist",
				warehouseId,
			)})
		} else {
			h.logger.Log(log.Error, err)
			h.response(w, http.StatusInternalServerError, api.GetWarehouseResponse{Error: "Failed to get warehouse"})
		}

		return
	}

	h.response(w, http.StatusOK, api.GetWarehouseResponse{Response: *warehouse})
```

siripr4 replied:
Changed this to use `if err != nil` syntax

olttwa replied:
I still see the `switch err` used in multiple places. Consider changing them to the `if` syntax. 
As I have elaborated above, switch, although convenient, tends to reduce readability, especially for the `nil` & `default` cases.

siripr4 replied:
Updated the syntax throughout the handler

olttwa commented on warehouse-management-service/internal/handler/handler.go:
```diff
@@ -469,7 +479,226 @@ func (h *handler) DeleteShelfBlock(w http.ResponseWriter, r *http.Request) {
+func (h *handler) CreateShelf(w http.ResponseWriter, r *http.Request) {
+	var createShelfRequest api.CreateShelfRequest
+
+	if r.Body == nil {
```
This check can be avoided. Error check at 532 should catch this & throw 400?

siripr4 replied:
The code on line [529](https://github.com/nilenso/siris-onboarding/pull/34/files/2fccb29993c28887cb5b75d65300dde6193018fb#diff-a3d7cbb31a8120c162b536abd4442e67f77b9a01f0666c4d80ed3f9c8b8f5e3cR529) panics when nil body is passed

olttwa replied:
Got it.

olttwa replied:
Using `json.Decoder` for large volumes of JSON data (in MBs) would be the correct decision.
However, `Decoder` isn't as straightforward as `Unmarshal`.
Checkout the malformed JSON explained in this blog to understand more details: https://dev.to/taqkarim/you-might-not-be-using-json-decoder-correctly-in-golang-12mb

I think we should stick to `Unmarshal` to keep things simple, and go to `Decoder` only when memory becomes a bottleneck for JSON parsing.

olttwa commented on warehouse-management-service/internal/handler/handler.go:
```diff
@@ -469,7 +479,226 @@ func (h *handler) DeleteShelfBlock(w http.ResponseWriter, r *http.Request) {
+		return
+	}
+
+	shelf := wms.NewShelf(
```
I think `CreateShelfRequest`, `UpdateShelfRequest` & `Shelf` can all be clubbed into 1 struct. It's ok to share a model-level struct at API level. If not very varying, 1 struct can have multiple tags too for JSON & SQL

siripr4 replied:
Updated this to use `Shelf` in the handler as well

olttwa commented on warehouse-management-service/internal/handler/handler.go:
```diff
@@ -469,7 +479,226 @@ func (h *handler) DeleteShelfBlock(w http.ResponseWriter, r *http.Request) {
+	}
+
+	err = h.shelfService.UpdateShelf(r.Context(), shelf)
+	switch err {
```
Same comment as above. `if err != nil` is preferable

siripr4 replied:
Changed this to use if err != nil syntax as well

olttwa commented on warehouse-management-service/internal/handler/handler.go:
```diff
@@ -469,7 +479,226 @@ func (h *handler) DeleteShelfBlock(w http.ResponseWriter, r *http.Request) {
+func (h *handler) UpdateShelf(w http.ResponseWriter, r *http.Request) {
+	var updateShelfRequest api.UpdateShelfRequest
+
+	if r.Body == nil {
```
Same comment as above. This check is redundant

olttwa commented on warehouse-management-service/pkg/database/postgres/init_test.go:
```diff
@@ -14,7 +14,8 @@ const (
 
 var warehouseService *WarehouseService
-var shelfService *ShelfBlockService
+var shelfBlockService *ShelfBlockService
```
`shelfBlockService` is confusing when used with `shelfService`. Consider renaming to something else like `block` or `shelves`? I lack context on the entire domain, so will leave the name to your judgement.

siripr4 replied:
A `shelfBlock` can be thought of as a rack that contains multiple shelves. I had considered naming it `block` however, it did not feel descriptive enough

olttwa commented on warehouse-management-service/pkg/database/postgres/shelf.go:
```diff
@@ -0,0 +1,198 @@
+var InvalidShelfBlock = errors.New("invalid shelfBlockId")
+
+func (s *ShelfService) GetShelfById(ctx context.Context, id string) (wms.Shelf, error) {
+	tx, err := s.db.BeginTx(ctx, nil)
```
Why are individual queries executed as part of a transaction?
If you did you do this for learning transactions, that's ok. Otherwise, there's no need for single queries to be part of a transaction.

olttwa commented on warehouse-management-service/pkg/database/postgres/shelf.go:
```diff
@@ -0,0 +1,198 @@
+}
+
+func (s *shelfQueriesImpl) createShelfTx(ctx context.Context, tx *sql.Tx, shelf wms.Shelf) error {
+	if shelfBlockExists, err := s.shelfBlockExistsTx(ctx, tx, shelf.ShelfBlockId); err != nil {
```
Can this logic can be moved from code to database by virtue of foreign keys?
In some high-scale applications, FKs are considered detrimental to performance. 
However, in this case, we can leverage them.

siripr4 replied:
The reasoning behind having this logic was to be able to distinguish between querying a non-existent record vs a generic internal error. However, this explicit check might be excessive and I am considering changing this to check for foreign key errors returned by the library.

olttwa commented on warehouse-management-service/pkg/database/postgres/shelf.go:
```diff
@@ -0,0 +1,198 @@
+}
+
+func (s *ShelfService) DeleteShelfById(ctx context.Context, id string) error {
+	tx, err := s.db.BeginTx(ctx, nil)
```
I think all CRUD operations can be moved from transactions to a single query.

olttwa replied:
I'm slightly confused since you reacted 👍 to above comment and I still see transactions being used.
If you intend to keep them, please resolve this conversation.

siripr4 replied:
I intend to refactor this once the CI/CD story I am working on currently is done, for the time being not changing this.

olttwa commented on warehouse-management-service/shelf.go:
```diff
@@ -0,0 +1,26 @@
+	"errors"
+)
+
+type Shelf struct {
```
As noted above, this struct can replace `CreateShelfRequest` & `UpdateShelfRequest`

siripr4 committed:
refactor: update error checking syntax in handler

Replace switch statements with if err != nil syntax to keep the success response in the same indentation level as the flow of the handler

siripr4 committed:
refactor: replace CreateShelfRequest and UpdateShelfRequest with Shelf

siripr4 committed:
refactor: rename package alias warehousemanagementservice -> wms

olttwa commented on warehouse-management-service/internal/handler/handler.go:
```diff
@@ -469,7 +479,212 @@ func (h *handler) DeleteShelfBlock(w http.ResponseWriter, r *http.Request) {
+		} else {
+			h.logger.Log(log.Error, err)
+			h.response(w, http.StatusInternalServerError, api.ShelfResponse{Error: "Failed to get shelf"})
+		}
```
`return` is missing at the end?

siripr4 replied:
Added the missing `return` statements

siripr4 committed:
fix: add missing return in handler

siripr4 committed:
refactor: switch -> if/else syntax in handler
